### ISSUE TITLE
Requiring unkown module problem

### ISSUE THREAD

**Issue by Eissek** [drapanjanas/re-natal#5]
Hi
I've been trying to use [react-native-vector-icons](https://github.com/oblador/react-native-vector-icons) in my re-natal generated project. But for some reason I keep on ending up with the requiring unknown module error.   
[<a href='http://postimg.org/image/nl0n2dqzx/' target='_blank'><img src='http://s30.postimg.org/nl0n2dqzx/2016_01_03_15_50_41_Genymotion_for_personal_use.jpg' border='0' alt="2016 01 03 15 50 41 Genymotion for personal use" /></a>](http://postimg.org/image/nl0n2dqzx/)
I've tried requiring the module using the following different paths below:     
`(def icon (js/require "react-native-vector-icons/FontAwesome.js"))`    
`(def icon (js/require "react-native-vector-icons/FontAwesome"))`     
`(def icon (js/require "FontAwesome.js"))`   
It seems to happen with other custom react native components as well. So it seems maybe I'm doing something wrong

**Comment by drapanjanas** [comment]
Hi
Thanks for reporting! Could you please check if the same happens when you run your app wthout figwheel: `lein prod-build` and refresh your app. I suspect it might be related to the fact that code loaded by figwheel bypass native packager.

**Comment by artemyarulin** [comment]
Yep, you cannot just require the package with figwheel (although it may work just fine for release build if you process your js using `react-native bundle` command). See more info how to get the package name in this post: http://blog.fikesfarm.com/posts/2015-07-24-using-react-native-components-in-clojurescript.html

**Comment by drapanjanas** [comment]
Thanks Artem! I have tested what is written and now its clear to me. One difference I found with current vesrion is that packager expanded the name 'react-native-swiper' into 'react-native-swiper/dist/index.js' - notice the js extention is there. 
So, to make it work with figwheel I had to:
1. Add a dummy call to require('react-native-swiper') to index.*.js file so that the module is packaged in index.*.bundle 
2. Use expanded name in cljs: (js/require "react-native-swiper/dist/index.js")
Well its a hack, but lets see if I can at least automate it somehow until better approach is discovered...

**Comment by artemyarulin** [comment]
I have an idea which I described here: https://github.com/artemyarulin/ktoa/issues/1

In short it's the same as you said but index.*.js should save required package inside a global variable which could be accessible for CLJS during development time and generate simple `js/require` call for the release build.

I haven't tried it yet - but looks like it's possible and there would be need to find a real package name

**Comment by drapanjanas** [comment]
Yes, I aslo thought of reusing the object from the initial call from index.\*.js. 
My thought was to build a map 'module name' -> result and then somehow shim require function in figwheel bridge code (I am not js expert, but maybe possible): first, check the map by module name, if not found fallback to original implementation.
The benefit would be that the cljs code would just use the normal `(js/require "my-module-name")` and the magic part would be only in the figwheel bridge code.

**Comment by artemyarulin** [comment]
Ha, cool idea, haven't though about it. Yeah, let's see who finds first free time to play with it :)

**Comment by Eissek** [comment]
Just to report back, yes using the lein prod-build without Figwheel worked just fine. Thanks for the help.

**Comment by drapanjanas** [comment]
As this is a significant limitation of using figwheel, let's keep this issue open until some working solution is implemented.

**Comment by decker405** [comment]
Since React Native packager transforms all `require`'s by injecting their module code somehow on each http request or bundle command, it makes it difficult to deal with require statements from Cljs. That's why it works when you use `lein prod-build`.

(I actually had to deal with react native trying to transform the require keyword [here](https://github.com/decker405/figwheel-react-native/blob/master/development.ios.js#L51-L52) - if you change that to a static string you'll find that it tells you a module is missing)

The easiest thing to do is to have an index.ios.js file, add any images or modules in there, and then access them from the Cljs code. For a more automated solution, I _think_ you could just shim the `require` function to do an http call for the module (see link below) and eval the result, similar to how [my figwheel shim code works](https://github.com/decker405/figwheel-react-native). In fact, you could probably just call the [myImportScripts function](https://github.com/decker405/figwheel-react-native/blob/master/development.ios.js#L89) with the correct path from inside the `require` shim. Just make sure you do it after you `require('react-native')`. Something like: ```
require = function(path) { 
    myImportScripts(path);
}``` after line 27? Actually, the `goog.require` function may just work now that I think about it...

Links: [packager documentation](https://github.com/facebook/react-native/tree/master/packager)

**Comment by drapanjanas** [comment]
Hi Will, 
First of all, man, I would like to personally thank you for creating and open sourcing the figwheel bridge! :)
My latest commit should solve the problem. I will give it some more testing before releasing, but I think it will do the trick. Later on, I will try to apply same pattern to images (now its a hacky macro) maybe I will scan images dir and require all images. But, steb by step.

**Comment by drapanjanas** [comment]
I have thought about fully automatic aproach like @decker405 have proposed, which would be the best solution. 
Though, I see two technical problems about it:
1. As I understand, the implementation of shimmed `require` function should be synchronous and return an object. I do not know how I can implement that with [myImportScripts function](https://github.com/decker405/figwheel-react-native/blob/master/development.ios.js#L89) or with `goog.require`
2. The parameter to `require` provided by cljs code has to be the path to js file like `react-native-swiper/dist/index.js` and not short `react-native-swiper`. I guess, it is not possible to resolve the path at runtime.

**Comment by drapanjanas** [comment]
Solved in v0.2.13

**Comment by sineer** [comment]
Hello friends!

Fyi, I still have this issue with exactly this "react-native-vector-icons/Ionicons" js/require.

This is with re-natal 0.2.41 react-native 0.35.0-rc0 and react-native-vector-icons 2.1.0

I can js/require "react-native-vector-icons" just fine but cannot get Ionicons to work no matter which variant of js/require I try it is always undefined just like 1st post in this thread...

I tried flushing packager cache, rebuild all, npm cache clean and reinstall after I delete node_modules... Nothing worked, in the end I modified react-native-vector-icons to export IonIcons seeing how I can js/require that but not the /Ionicons.js from cljs (strange! I wonder why).

I use half a dozen other RN libs with js/require without issue, just that one IonIcons es6 with no export default maybe I dunno...
